pkg/api/data.go

```diff
@@ -0,0 +1,152 @@
+package api
+
+import (
+	"github.com/grafana/grafana/pkg/api/dtos"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/plugins"
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/grafana/grafana/pkg/tsdb"
+)
+
+// QueryData endpoint for querying backend data source plugins.
+// POST /api/ds/query
+func (hs *HTTPServer) QueryData(c *models.ReqContext, reqDto dtos.QueryDataRequest) Response {
+	if len(reqDto.Queries) == 0 {
+		return Error(500, "No queries found in query", nil)
+	}
+
+	request := &tsdb.TsdbQuery{
+		User: c.SignedInUser,
+	}
+
+	var ds *models.DataSource
+	var err error
+	for i, query := range reqDto.Queries {
+		if i == 0 {
+			request.TimeRange = tsdb.NewTimeRange(query.From, query.To)
+			ds, err = hs.DatasourceCache.GetDatasource(query.DatasourceID, c.SignedInUser, c.SkipCache)
+			if err != nil {
+				if err == models.ErrDataSourceAccessDenied {
+					return Error(403, "Access denied to datasource", err)
+				}
+				return Error(500, "Unable to load datasource meta data", err)
```

Comment:
Metadata should be a single word:
```suggestion
				return Error(500, "Unable to load datasource metadata", err)
```